tests/test_bert_embedder.py

```diff
@@ -0,0 +1,232 @@
+            {"tokens": TextField(tokens2, {"bert": self.token_indexer})}
+        )
+
+        batch = Batch([instance1, instance2])
+        batch.index_instances(self.vocab)
+
+        padding_lengths = batch.get_padding_lengths()
+        tensor_dict = batch.as_tensor_dict(padding_lengths)
+        tokens = tensor_dict["tokens"]
+
+        assert tokens["bert"].tolist() == [
+            [
+                1103,
+                3613,
+                2556,
+                3613,
+                3058,
+                17594,
+                4874,
+                1166,
+                1103,
+                16688,
+                3676,
+                0,
+                0,
+            ],
+            [
+                1103,
+                3613,
+                3058,
+                17594,
+                4874,
+                1166,
+                1103,
+                2495,
+                15039,
+                2050,
+                16688,
+                8468,
+                3702,
+            ],
+        ]
+
+        assert tokens["bert-offsets"].tolist() == [
+            [0, 1, 3, 4, 5, 6, 7, 8, 9, 10],
+            [0, 1, 2, 3, 4, 5, 6, 7, 10, 11],
+        ]
+
+        token_embedder = PretrainedBertEmbedder(
+            self.model_name,
+            requires_grad=False,
+            top_layer_only=False,
+            special_tokens_fix=0,
+        )
+
+        bert_vectors = token_embedder(tokens["bert"])
+        assert list(bert_vectors.shape) == [2, 13, 768]
+
+        # Offsets, should get 10 vectors back.
+        bert_vectors = token_embedder(
+            tokens["bert"], offsets=tokens["bert-offsets"]
+        )
+        assert list(bert_vectors.shape) == [2, 10, 768]
+
+        # Now try top_layer_only = True
+        tlo_embedder = PretrainedBertEmbedder(
+            self.model_name, top_layer_only=True, special_tokens_fix=1
+        )
+        bert_vectors = tlo_embedder(tokens["bert"])
+        assert list(bert_vectors.shape) == [2, 13, 768]
+
+        bert_vectors = tlo_embedder(
+            tokens["bert"], offsets=tokens["bert-offsets"]
+        )
+        assert list(bert_vectors.shape) == [2, 10, 768]
+
+    def test_max_length(self):
+        """Test that max input length works."""
+
+        token_embedder = PretrainedBertEmbedder(
+            self.model_name,
+            requires_grad=False,
+            top_layer_only=True,
+            special_tokens_fix=0,
+        )
+
+        sentence = "the " * 512
+        tokens = [Token(word) for word in sentence.split()]
+
+        instance = Instance(
+            {"tokens": TextField(tokens, {"bert": self.token_indexer})}
+        )
+
+        batch = Batch([instance])
+        batch.index_instances(self.vocab)
+
+        padding_lengths = batch.get_padding_lengths()
+        tensor_dict = batch.as_tensor_dict(padding_lengths)
+        tokens = tensor_dict["tokens"]
+        token_embedder(tokens["bert"], tokens["bert-offsets"])
```

Comment:
Im learning the pytest, just wondering should an assertion be added here?

Reply:
This test basically checks that a sentence with `max_len` tokens i.e. 512, does not raise an exception and passes. Let me see if I can make this clearer. From what I remember, an alternative was to put this in a `try/except` block.

Reply:
```suggestion
        try:
            token_embedder(tokens["bert"], tokens["bert-offsets"])
        except Exception as exception:
            assert False, f"Unexpected exception raised: {exception}"
```

So I tried the above code, but I'm not sure I prefer this because it returns an `AssertionError`, instead of the original `IndexError`.  The output looks like this:

```
FAILED tests/test_bert_embedder.py::TestPretrainedBertEmbedder::test_max_length - AssertionError: Unexpected exception raised: index 0 is out of bounds for dimension 1 with size 0
```

What do you think?

Reply:
You need to use whatever the equivalent of `self.assertRaises` from unittest.TestCase is in pytest.

Reply:
Yes, I have used that in the next test that is supposed to raise an exception. This test should simply pass. I am deliberately making it fail to get the above output.

Reply:
But what's the actual test in this function?

Reply:
In that case, the test explicitly needs to check that no error was raised.

Reply:
Let me play around with this.